astropy/cosmology/funcs.py

```diff
@@ -253,3 +253,47 @@ def f(z):
         raise CosmologyError(f"Best guess z={res.x} is very close to the lower z limit {zmin}.\n"
                              "Try re-running with a different zmin.")
     return res.x
+
+
+def z_at_value(func, fval, zmin=1e-8, zmax=1000, ztol=1e-8, maxfun=500,
+               method='Brent', bracket=None, verbose=False):
+    """``scalar_z_at_value`` for vectorized inputs. Docstring set below."""
+    # `fval` can be a Quantity, which isn't (yet) compatible w/ `numpy.nditer`
+    # so we strip it of units for broadcasting and restore the units when
+    # passing the elements to `scalar_z_at_value`.
+    unit = getattr(fval, 'unit', 1)  # can be unitless
+
+    # bracket must be an object array (assumed to be correct) or a 'scalar'
+    # bracket: 2 or 3 elt sequence
+    if not isinstance(bracket, np.ndarray):  # 'scalar' bracket
+        if bracket is not None and len(bracket) not in (2, 3):
+            raise ValueError("`bracket` is not an object array "
+                             "nor a 2 (or 3) element sequence.")
+        else:  # munge bracket into a 1-elt object array
+            bracket = np.array([bracket, ()], dtype=object)[:1]
+    if bracket.dtype != np.object_:
+        raise TypeError(f"`bracket` has dtype {bracket.dtype}, not 'O'")
+
+    # make multi-dimensional iterator for all but `method`, `verbose`
+    # TODO: figure out 'reduce_ok', so scalar returns scalar
+    with np.nditer(
+        [np.asanyarray(fval), zmin, zmax, ztol, maxfun, bracket, None],
+        flags = ['refs_ok'],
+        op_flags = [*[['readonly']] * 6,  # ← inputs  output ↓
+                    ['writeonly', 'allocate', 'no_subtype']],
+        casting="no",
+    ) as it:
+        for fv, zmn, zmx, zt, mfe, brkt, zs in it:  # ← elt-wise unpack and eval ↓
+            zs[...] = scalar_z_at_value(func, fv * unit, zmin=zmn, zmax=zmx,
+                                        ztol=zt, maxfun=mfe, bracket=brkt.item(),
+                                        # not broadcasted
+                                        method=method, verbose=verbose)
+        # since bracket is an object array, the output will be too, so it is
```

Comment:
Not sure any more, but I thought you can give the dtype using the `op_dtypes` argument for `nditer` - object would just be its best guess.

Reply:
One can give the dtype. I've lightly tested it and it's worth another look, but I was worried about changing the dtype from e.g. float64 to float32. the `op_dtypes` argument appears to need a dtype for every argument... maybe there's a way around that.

Reply:
I'm pretty sure you can give `None` to indicate just use the one that the array has.

Reply:
Unfortunately, it appears not. That would be a nice extension.

<img width="867" alt="Screen Shot 2021-07-13 at 17 29 31" src="https://user-images.githubusercontent.com/8949649/125527729-9d486e4a-518e-44bd-aa50-154315f64302.png">

Reply:
should be `op_dtypes = [None]*6+[fval.dtype]`, I think (or just a missing `,` after `None`

Reply:
I should have written that as ``[None]`` for it to work. 🤦